docs/current_docs/developer-guide/go/742930-services.mdx

```diff
@@ -0,0 +1,260 @@
+
+One use case is for testing, where you need to be able to spin up ephemeral databases to run tests against. You might also use this to access a web UI in a browser on your desktop.
+
+Here is another example call for the Dagger Function shown previously, this time exposing the HTTP service on host port 9000:
+
+```shell
+dagger call http-service up --ports 9000:8080
+```
+
+This service can then be accessed by clients on the host. Here's an example:
+
+```shell
+curl localhost:9000
+```
+
+The result will be:
+
+```shell
+Hello, world!
+```
+
+:::note
+By default, each service port maps to the same port on the host. To specify a different mapping, use the additional `--ports` argument to `dagger call...up` with a list of host/service port mappings. To bind ports randomly, use the `--random` argument.
+:::
+
+## Expose host services to functions
+
+Dagger Functions can also receive host services as function arguments of type `Service`, in the form `tcp://<host>:<port>`. This enables client containers in Dagger Functions to communicate with services running on the host.
+
+:::note
+This implies that a service is already listening on a port on the host, out-of-band of Dagger.
+:::
+
+Here is an example of how a container running in a Dagger Function can access and query a MariaDB database service (bound using the alias `db`) running on the host.
+
+```go file=./snippets/services/expose-host-services-to-dagger/main.go
+```
+
+Before calling this Dagger Function, use the following command to start a MariaDB database service on the host:
+
+```shell
+docker run --rm --detach -p 3306:3306 --name my-mariadb --env MARIADB_ROOT_PASSWORD=secret  mariadb:10.11.2
+```
+
+Here is an example call for this Dagger Function:
+
+```shell
+dagger call user-list --host-service=tcp://localhost:3306
+```
+
+The result will be:
+
+```shell
+Host    User
+%       root
+localhost       mariadb.sys
+localhost       root
+```
+
+## Persist service state
+
+Dagger cancels each service run after a 10 second grace period to avoid frequent restarts. To avoid relying on the grace period, use a cache volume to persist a service's data, as in the following example:
+
+```go file=./snippets/services/persist-service-state/main.go
+```
+
+This function uses Redis's SAVE command to ensure data is synced. By default, Redis flushes data to disk periodically.
+
+Here is an example call for this Dagger Function:
+
+```shell
+dagger call redis-service
+```
+
+The result will be:
+
+```shell
+abc
+```
+
+## Start and stop services
+
+Services are designed to be expressed as a Directed Acyclic Graph (DAG) with explicit bindings allowing services to be started lazily, just like every other DAG node. But sometimes, you may need to explicitly manage the lifecycle in a Dagger Function.
+
+Here's an example which demonstrates explicitly starting a Docker daemon for use in a test suite:
+
+```go file=./snippets/services/start-stop-services/main.go
```

Comment:
This example may need to be rewritten. I could not test it. Can you help @vito?

Reply:
Yeah, I think building on the Redis example here is a good idea, since it's a natural progression: if you want to avoid the 10-second grace period, you can explicitly start Redis which will keep it running until you stop it. Also probably helps to avoid avoid "Docker in Dagger" which is kind of a complex topic and probably draws too much attention away from the task at hand. I can update the example for this.